lib/portlayer/vspc/handlers.go

```diff
@@ -0,0 +1,195 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vspc
+
+import (
+	"bytes"
+	"crypto/rand"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/vmware/vic/pkg/telnet"
+)
+
+// Handler is the handler struct for the vspc
+type Handler struct {
+	vspc *Vspc
+}
+
+// DataHdlr is the telnet data handler
+func (h *Handler) DataHdlr(w io.Writer, b []byte, tc *telnet.Conn) {
+	vm, exists := h.vspc.vmFromTelnetConn(tc)
+	if !exists {
+		log.Errorf("vspc datahandler cannot find the vm with the provided data")
+	}
+	vm.remoteConn.Write(b)
+}
+
+// CmdHdlr is the telnet command handler
+func (h *Handler) CmdHdlr(w io.Writer, b []byte, tc *telnet.Conn) {
+	if isKnownSuboptions(b) {
+		log.Infof("vspc received KNOWN-SUBOPTIONS command")
+		h.HandleKnownSuboptions(w, b)
+	} else if isDoProxy(b) {
+		log.Infof("vspc received DO-PROXY command")
+		h.HandleDoProxy(w, b)
+	} else if isVmotionBegin(b) {
+		log.Infof("vspc received VMOTION-BEGIN command")
+		h.HandleVmotionBegin(w, tc, b)
+	} else if isVmotionPeer(b) {
+		log.Infof("vspc received VMOTION-PEER command")
+		h.HandleVmotionPeer(w, b)
+	} else if isVMUUID(b) {
+		log.Infof("vspc received VMUUID command")
+		h.HandleVMUUID(w, tc, b)
+	} else if isVmotionComplete(b) {
+		log.Infof("vspc received VMOTION-COMPLETE command")
+		h.HandleVmotionComplete(w, tc, b)
+	} else if isVmotionAbort(b) {
+		log.Infof("vspc received VMOTION-ABORT command")
+		h.HandleVmotionAbort(w, b)
+	}
+}
+
+// HandleVMUUID handles the telnet vm-uuid response
+func (h *Handler) HandleVMUUID(w io.Writer, tc *telnet.Conn, b []byte) {
+	vmuuid := strings.Replace(string(b[3:len(b)-1]), " ", "", -1)
+	log.Infof("vmuuid of the connected containerVM: %s", vmuuid)
+	// check if there exists another vm with the same vmuuid
+	vm, exists := h.vspc.getVM(vmuuid)
+	if !exists {
+		// create a new vm associated with this telnet connection
+		vm = NewVM(tc)
+		log.Infof("attempting to connect to the attach server")
+		remoteConn, err := net.Dial("tcp", fmt.Sprintf("%s:%d", h.vspc.attachSrvrAddr, h.vspc.attachSrvrPort))
+		if err != nil {
+			log.Errorf("cannot connect to the attach-server: %v", err)
+		}
+		vm.remoteConn = remoteConn
+		vm.vmUUID = vmuuid
+		h.vspc.addVM(vmuuid, vm)
+		// relay Reads from the remote system connection to the telnet connection associated with this vm
+		go h.vspc.relayReads(vm, remoteConn)
```

Comment:
I don't agree with the writer that ioutil.ReadAll is to be avoided at all costs :). I guess it is only bad if the user does not really know what concrete type the underlying io.Reader which is not the case here. That said, I looked into the bytes package and there is a more efficient way to read the buffer content than ioutil.ReadAll by using buffer.Bytes() and the buffer.Reset()